leakcanary-android-proguard/src/main/java/com/squareup/leakcanary/proguard/LeakcanaryProguardPlugin.kt

```diff
@@ -0,0 +1,42 @@
+package com.squareup.leakcanary.proguard
+
+import com.android.build.gradle.AppExtension
+import org.gradle.api.Plugin
+import org.gradle.api.Project
+
+class LeakcanaryProguardPlugin : Plugin<Project> {
+  override fun apply(project: Project) {
+    val extension = project.extensions.create("leakCanary", LeakCanaryProguardExtension::class.java)
+
+    val androidPluginHandler = { _: Plugin<*> ->
+      val android = project.extensions.getByType(AppExtension::class.java)
+
+      project.afterEvaluate {
+        android.applicationVariants.all { variant ->
+
+          if (extension.obfuscatedVariantNames?.contains(variant.name) == true) {
+            val copyProguardMappingFileTask = project.tasks.create(
+                "leakCanaryCopyProguardMappingFor${variant.name.capitalize()}",
+                CopyProguardMappingFileTask::class.java
+            ) {
+              it.variantDirName = variant.dirName
+              it.mappingFile = variant.mappingFile
+              it.mergeAssetsDirectory = variant.mergeAssetsProvider.get().outputDir.get().asFile
```

Comment:
Due to Task Configuration Avoidance, you want to make sure that this takes place when the task is executed, not when it is configured or created. The call to provider get() will create the task and you want this to happen during execution, not during evaluation. You probably need either a custom task or a method override in the inline task class.

Reply:
I have to read about Task Configuration Avoidance, do you know if there are any open source plugins that implement it already? I would like to take a look at the example implementation :)

Reply:
There are a few but all implementations are very specific and it would be hard to find one that has the potential same issues. Here you just want to avoid one thing: realize / create the task `mergeAssets` too soon = when your task is configured. It happens when you call `taskProvider.get()`.

To avoid this, you need to make this call during a task action and not during the config block of the task. The most clean approach is to create your own task class, have a method with the `@TaskAction` annotaton and put the logic there. Then instantiate your task class. You have already done that in the plugin.

Reply:
Thank you, I know where to start now :)

Reply:
I checked the plugin source code and yes: a variant should be cacheable, as it is serializable.. fingers crossed.
That's also something you can test..
So, just pass the variant, and extract the various properties from it from inside the task class itself, from within the `TaskAction` annotated method. 
The only alternative I see is to replicate the code logic to extract the location of the output of the mergeAsset Task.
But it doesn't look like a good idea.

Reply:
There are indeed a few things to say here. I could talk with some people from gradle:
* we do the `mergeAssetsProvider.get()` call in the configuration lambda of the plugin, as we are using the `register` api, then it will not be that bad: at this point we are sure the `mergeAssets` task is gonna be created.
* if we want to be more explicit, we can pass the variant as an input to the new task, yes it should be cacheable still as variant are serializable.
* if we want to be more classy, we could use properties: https://docs.gradle.org/current/userguide/lazy_configuration.html

Both ways will be good from a performance perspective.